### PR TITLE
[autoscaler] Event summarizer reports launch failure

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Partly addresses https://github.com/ray-project/ray/issues/20774 by registering node launcher failures in driver logs, via the event summarizer.
This way, users can at least tell that the launch failed from the driver logs.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/event_summarizer.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from typing import Any, Callable, Dict, List
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/event_summarizer.py:1]: Nothing changed in this file besides sticking methods behind a lock for thread safety.

**Review by sasha-s — APPROVED** [review]

**Review by wuisawesome — COMMENTED** [review]
Are you able to repro this issue/share the output on a real cluster? 

On a somewhat related note, this message will be repeated a bunch on the driver (since the autoscaler should retry every iteration) right? I guess that's not a regression though.

**Comment by DmitriGekhtman** [comment]
> Are you able to repro this issue/share the output on a real cluster?

Will do -- should be possible to repro by putting garbage in the worker node config.

**Comment by DmitriGekhtman** [comment]
> On a somewhat related note, this message will be repeated a bunch on the driver (since the autoscaler should retry every iteration) right? I guess that's not a regression though.

Let's think later about how best to rate limit in these logs. https://github.com/ray-project/ray/issues/20855

**Comment by DmitriGekhtman** [comment]
@wuisawesome
Here's what it looks like on AWS 
Screen Shot 2021-12-06 at 3 13 39 PM

**Review by wuisawesome — APPROVED** [review]

**Comment by wuisawesome** [comment]
Looks good! (we should probably de-dupe these messages, but this is much more informative than before)

**Comment by DmitriGekhtman** [comment]
> we should probably de-dupe these messages
I agree. Though in some sense, they are not duplicated -- the autoscaler really did repeatedly try and fail to launch two nodes.

**Comment by wuisawesome** [comment]
Yup agreed, also to be clear I mean we should _eventually_ do it, not in this PR

**Comment by AmeerHajAli** [comment]
Lets make sure to propagate the exception to driver as discussed.

**Comment by DmitriGekhtman** [comment]
> Lets make sure to propagate the exception to driver as discussed.

Come to think of it, the `add_once_per_interval` method Eric added to the event summarizer should serve nicely for this (to avoid spamming with the traceback every 5 seconds.)

**Comment by DmitriGekhtman** [comment]
A'ight. Not particularly aesthetic, but gets the traceback out without spamming it every 5 seconds:
Screen Shot 2021-12-06 at 8 56 34 PM

**Comment by DmitriGekhtman** [comment]
Green, merging.

**Pull request merged**